### DISCUSSION TITLE
Deploy on AWS Lambda

### DISCUSSION THREAD

**Discussion by semsphy** [osm-search/Nominatim#3601] [Q&A]
Hi Guys,

## Description

I attempted to deploy `Nominatim` after converting database to SQLite.
I’m using [`lambda-adapter`](https://github.com/awslabs/aws-lambda-web-adapter) to proxy requests to the internal web server. 

When I tested the build locally, it worked as expected. However, after deploying it to an AWS Lambda URL, I encountered the error message in `Error` part. 

As I googled around, it seems that AWS Lambda cannot handle the asynchronous API call in `nominatim_api.server.falcon.server:run_wsgi`

Additionally, I tried using `CMD ["gunicorn", "-b=:8080", "-w=1", "nominatim_api.server.falcon.server:run_wsgi"]`, but it still didn’t work because it is used for synchronous workers.

## Question

I would like to extend the server to support synchronous workers. 
- Do you have any advice?
- Alternatively, does anyone have suggestions for improving the CMD configuration?

## Error
![image](https://github.com/user-attachments/assets/158f8ed0-6ef3-4ec8-ad0d-1d2103691c7f)

## Dockerfile sample 

```Dockerfile
# Dockerfile
#...

COPY data.sqlite /app/data.sqlite
COPY env /app/.env

RUN sqlite3 /app/data.sqlite "\
    PRAGMA journal_mode = WAL; \
    PRAGMA synchronous = NORMAL; \
    PRAGMA cache_size = -2000000; \
    PRAGMA temp_store = MEMORY; \
    PRAGMA mmap_size = 30000000000; \
    VACUUM; \
    ANALYZE;"

# Clean up build artifacts and temporary files
RUN apt-get -y remove --purge --auto-remove build-essential \
    && rm -rf /tmp/* /var/tmp/* \
    && pip cache purge \
    && apt-get clean \
    && rm -rf /var/lib/apt/lists/*

FROM scratch

COPY --from=build / /

COPY --from=public.ecr.aws/awsguru/aws-lambda-adapter:0.8.4 /lambda-adapter /opt/extensions/lambda-adapter

ENV PROJECT_DIR=/app

ARG USER_AGENT
ENV USER_AGENT=${USER_AGENT}


WORKDIR /app

CMD ["gunicorn", \
     "--bind", ":8080", \
     "--enable-stdio-inheritance", \
     "--workers", "1", \
     "--worker-class", "uvicorn.workers.UvicornWorker", \
     "nominatim_api.server.falcon.server:run_wsgi"]
```

**Comment by lonvia** [comment]
This might be possible by using a similar approach as the [synchronous API wrapper](https://github.com/osm-search/Nominatim/blob/master/src/nominatim_api/core.py#L406): wrap each endpoint into `run_until_complete()` function to hide the asynchronous functions.

You basically need to extend the [falcon server script](https://github.com/osm-search/Nominatim/blob/master/src/nominatim_api/server/falcon/server.py) with synchronous functions for the middleware and do the `run_until_complete()` wrapping in the (EndPoint wrapper)(https://github.com/osm-search/Nominatim/blob/master/src/nominatim_api/server/falcon/server.py#L113).

If it is not too much code duplication, a version that supports both might even be something that can go into main line.
